### ISSUE TITLE
[$1000] Attachment - No feedback or error message when choosing invalid image file to be sent on chat

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19728]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the Expensify app on the web browser
2. Go to any chat conversation
3. Click on the "+" icon(actions button)  of the chat
4. Click on Add attachment
5. Select an invalid or corrupted image file from the file explorer(you can create invalid image file by changing any .txt or .pdf file to .jpg simply or you can find sample invalid image attached in thread)
6. Observe the infinite loading

## Expected Result:
The app should display a feedback or error message indicating that the file is not valid or just upload and send as a file without trying to display preview(if the image file is invalid)

## Actual Result:
The app does not show any feedback or error message and tries to preview the image file with infinite loading

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.19.2

**Reproducible in staging?:** yes

**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/8e276d99-a196-4816-897d-be509bff7911

https://github.com/Expensify/App/assets/93399543/7f384736-400a-4265-9e89-c5023aa5e49e

**Expensify/Expensify Issue URL:**

**Issue reported by:** @kalll

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1685001824882209

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01764cde9077e5564b</li>
        <li>Upwork Job ID: 1664297284350144512</li>
        <li>Last Price Increase: 2023-07-14</li>
    </ul>
</details>

**Comment by sonialiap** [comment]
👍 adding External label

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When attach an invalid image file, the AttachPreview shows an infinite loading without showing any error

### What is the root cause of that problem?

We're using the Image component defined [here](https://github.com/Expensify/App/blob/main/src/components/Image/index.js) in our [ImageView](https://github.com/Expensify/App/blob/main/src/components/ImageView/index.js) component. 

The root cause of this issue is that we don't use `onError` callback of the Image component in `ImageView`
https://github.com/Expensify/App/blob/d661708d0aae864caeceb98801e921f07c0954a8/src/components/ImageView/index.js#L244-L254

https://github.com/Expensify/App/blob/d661708d0aae864caeceb98801e921f07c0954a8/src/components/ImageView/index.js#L283-L290

In addition, we didn't explicitly add `onError` props to its propTypes [here](https://github.com/Expensify/App/blob/main/src/components/Image/imagePropTypes.js)


### What changes do you think we should make in order to solve the problem?

We have to add `onError` props to the `Image` component explicitly and use it in our `ImageView` component. In `onError` callback, we need to do something like

```
    this.setState({isLoading: false, loadingError: 'Failed to load image'});
```

and show error message instead of loading indicator.

<details>
    <summary>Result</summary>

https://github.com/Expensify/App/assets/126839717/789bc941-6a98-47ab-9cb8-f47620c943cf


</details>

### What alternative solutions did you explore? (Optional)

**Comment by marcaaron** [comment]
@aimane-chnaif can you take a look at the proposals so far?

**Comment by s-alves10** [comment]
@aimane-chnaif 
We're using `Image` component of react-native and I think we should use error handling of the component. This is good for consistency as well.

What if a file is marked valid in our engine but can't be shown in `Image` view? Or inverse case?
We can show error message for `Image` can't render: `Invalid or not supported image format`

**Comment by marcaaron** [comment]
@aimane-chnaif can you summarize what you think we should do for this one? When you believe we're ready to move forward with something I will share my thoughts on that proposal, thanks!

**Comment by KALLL** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The bug is that when the user attaches an image file that is invalid or corrupted, the app tries to display a preview of the image and takes forever. The expected behavior is that the app should display a feedback or error message indicating that the file is not valid, or just upload and send the file as a file without trying to display a preview.
### What is the root cause of that problem?
The root cause of the problem is we are not handling onError well on the 
https://github.com/Expensify/App/blob/d661708d0aae864caeceb98801e921f07c0954a8/src/components/ImageView/index.js#L244-L254
### What changes do you think we should make in order to solve the problem?


- Import the icon.png file

`import icon from '../images/placeholder-image.png';`

- Define a state variable that will store the real file
```
this.state = {
  realFile: null
}
```
- Define a function that will handle the file selection
```
handleFileSelect = (event) => {
  // Get the selected file from the event
  const file = event.target.files[0];
  // Set the realFile state to the selected file
  this.setState({realFile: file});
  // Set the image source to the file URL
  this.setState({imageSource: {uri: URL.createObjectURL(file)}});
}
```
- Define a function that will handle the image error
```
handleImageError = () => {
  // Set the image source to placeholder-image.png
  this.setState({imageSource: icon});
}
```
- Modify the function that will handle the file upload to assign realFile
```
handleFileUpload = () => {
  // Get the real file from the state
  const file = this.state.realFile;
  // Check if the file is not null
  if (file) {
    // Upload the file using your preferred method
}    
  }
```
`<input type="file" onChange={this.handleFileSelect} />`
### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by KALLL** [comment]
So from my understanding and research from different popular platforms, here is what should be done. We'll check if the file is image type and if it is valid. If valid, preview will be shown. If not valid "invalid image" saying placeholder image will be previewed and when send is clicked it should send the file as it is with thumbnail placeholder "invalid image". This way it will get consistent. 

Here is a sample from slack for such attachments.
The first placeholder is for invalid image and the second for valid attachment
![image](https://github.com/Expensify/App/assets/34364253/52bb1088-d814-4146-bfc9-88e0c97c8dbc)

I proposed rough implementation of this at https://github.com/Expensify/App/issues/19728#issuecomment-1585813977 . I can provide efficient and specific code if the approach is liked.

**Comment by KALLL** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Infinite loading when a corrupted image is uploaded
### What is the root cause of that problem?
In ImageView, we do not handle the `onError` callback that react native Image component provides. This callback is invoked when a corrupted file is uploaded. In such a scenario, the `isLoading` state remains true because we do not handle the error cases
### What changes do you think we should make in order to solve the problem?
We should add an `onError` callback to the `<Image/>` tag and on error we can display the error.
```
<Image
    onError={this.imageLoadingError}
    // ....
/>
```
and listen to the errors callback
```
imageLoadingError({nativeEvent}) {
    this.setState({isLoading: false, imageHasError: true });
}
```
then show the error in the modal just like we do for a pdf

```
// add this as the first code block in render method
if (this.state.imageHasError && !this.state.isLoading) {
    return (<View style={[styles.imageViewContainer, styles.overflowHidden]}>
        <Text style={[styles.textLabel, styles.textLarge]}
            {this.props.translate('attachmentView.failedToLoadImage')}
        </Text>
    </View>)
}
```
To display the error message `attachmentView.failedToLoadImage`, we need to add language keys for it. We also need to reset `imageHasError` when we reload the image (for example, on a component update).

I recommend either disabling or removing the send button or using a close button instead in the attachment modal. The PDFView button should also be disabled for a broken pdf to prevent the user from sending it and causing a report error.

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by Christinadobrzyn** [comment]
Hey @aimane-chnaif could you check this GH and see if it's the same root issue? https://github.com/Expensify/App/issues/20632

**Comment by KALLL** [comment]
@aimane-chnaif @sonialiap @marcaaron Is there anyone's proposal chosen here? or we're still waiting?

**Comment by marcaaron** [comment]
I would advise we bring the conversation to Slack as the default way to unblock this or any other issue. If you think it should be internal then ask the BZ member to add the `Internal` label. The spirit of the recent changes [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1686668655175609) is to avoid pinging Expensify engineers in the issues unless there is some clear proposal to present that you feel is likely going to get approved without much additional comment.

**Comment by KALLL** [comment]
> @KALLL you have copied my exact proposal from [here](https://github.com/Expensify/App/issues/20632#issuecomment-1588280509)

Yes, sorry for not mentioning your name I just can't find a space for that. I posted your proposal here because this issue is taking more than 2 weeks and found your proposal helpful, but I'd rather just link it.

**Comment by sonialiap** [comment]
@aimane-chnaif following Mark's response, do you think this issue should be brought up in slack? If yes, please link it to this GH issue

**Comment by sonialiap** [comment]
@aimane-chnaif do we have a proposal that matches our desired solution? Or one that's pretty close and just needs to be tweaked a little bit?

**Comment by sonialiap** [comment]
Ok, I'll move this to internal

**Comment by michaelhaxhiu** [comment]
🧇 I'm removing AlexM's assignment and triggering assignment of a new CME.

**Comment by MariaHCD** [comment]
Anyone able to give me an overview of the current status of this issue? :) Maybe @aimane-chnaif?

**Comment by MariaHCD** [comment]
We're continuing the discussion in the slack conversation: https://expensify.slack.com/archives/C01GTK53T8Q/p1686912137101449

**Comment by MariaHCD** [comment]
@aimane-chnaif Let's start accepting proposals to block the user if the file extension and the actual format of the file are different or if there is no file extension at all.

Backend fix is here: https://github.com/Expensify/Web-Expensify/pull/38011

**Comment by MaciejSWM** [comment]
Hey! I’m Maciej from Software Mansion, an expert agency, and I’d like to investigate this issue.

**Comment by MaciejSWM** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

1. No error message presented when uploaded file's extension is different from extension derived from file's headers.
2. Infinite loading spinner is shown when file is invalid.

### What is the root cause of that problem?

Extension checking is currently done only based on file name and doesn't catch mismatches on file header vs file extension.

### What changes do you think we should make in order to solve the problem?

1. Read the file's headers before upload, then validate if type in the headers matches file's extension. If not, then show error message to user.
2. If there are any problems with reading the file on the frontend side, also show error message to the user. (Reading the file on the frontend side will get rid of the need to add any `onerror` checks in `<Image>`, as any errors will be detected during reading the image.

**Comment by MaciejSWM** [comment]
@aimane-chnaif Mapping magic numbers to file extensions sounds good. Other possibility is to read media type (e.g. `image/webp` like this:
```
const fileReader = new FileReader;
fileReader.readAsDataURL(file);
// data:image/webp;base64,UklGRoQQAQBXRUJQVlA4WAoAAAA…cAAAALiPgABKAVfL+fHBQAB3yFLAAAAAAAAAAAAAAAAAAAA==
```

**Comment by MaciejSWM** [comment]
But overally @allroundexperts's alternative proposal is in line with the newest requirements for this issue:

> @aimane-chnaif Let's start accepting proposals to block the user if the file extension and the actual format of the file are different or if there is no file extension at all.

And checking magic numbers seems like a better idea that checking media types, as media types are generally meant for browsers only, while magic numbers seem like the more proper way to do it. I would accept @allroundexperts's alternative proposal. @aimane-chnaif LMK if you agree.

**Comment by MariaHCD** [comment]
> do you mind sharing backend logic of getting correct file type? i.e. what library used?

We're not using any library in the backend. We get the mime type of the uploaded file (if the browser provided this information). And then execute the command `file -b --mime` to get the actual file type. If the uploaded file type != the actual file type, we thrown an exception.

**Comment by MaciejSWM** [comment]
To sum up the discussion, I can see we need two changes:
- `onError` handler when error is thrown in `<ImageView />` - it will handle anything that is not handled by mime-checking
- Mime type checking solution that will be similar to `file -b --mime`

Those requirements are in line with @allroundexperts's proposal, so I think it should be accepted.

@aimane-chnaif waiting for your final decision.

**Comment by MaciejSWM** [comment]
@aimane-chnaif if the type checking shall be as similar to the checks on the backen side as possible, using [mime-type mappings](https://gist.github.com/AshHeskes/6038140) might be better that using magic headers.

**Comment by MaciejSWM** [comment]
Oh, to obtain the mime type, you can [see this](https://github.com/Expensify/App/issues/19728#issuecomment-1617930337). The mime-type will be right after `data:`, as per [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs).

**Comment by MaciejSWM** [comment]
@aimane-chnaif  I used code below to examine a couple of different file types, here are the findings:

Code:
```
const fileReader = new FileReader;
fileReader.onloadend = () => {
    console.log('dataUrl', fileReader.result)
}
fileReader.readAsDataURL(_file);
```

Extracted mime types:
- `.csv` file -> `text/csv`
- `.doc` file -> `application/msword`
- `.docx` file -> `application/vnd.openxmlformats-officedocument.wordprocessingml.document`
- `.pdf` file -> `application/octet-stream` or `application/pdf`
- `.exe` -> `application/x-msdownload`
- `.xls` -> `application/vnd.ms-excel`
- `.xslx` -> `application/vnd.openxmlformats-officedocument.spreadsheetml.sheet`
- `.dmg` -> `application/octet-stream`
- `.zip` -> `application/zip`

The findings are the following:
- `application/x-msdownload` was the only mime type not listed in IANA website, all the others were present [here](https://www.iana.org/assignments/media-types/media-types.xhtml).
- Some files on my disk (`.pdf`, `.dmg`) had no specific mime type, instead they returned `application/octet-stream`. In that case I was still able to open them correctly on my Mac, I assume Mac inferred the type from the file extension to do it. As per IANA website: "The "octet-stream" subtype is used to indicate that a body contains
arbitrary binary data.  The set of currently defined parameters is:" ([link](https://www.iana.org/assignments/media-types/application/octet-stream))
- Some mime types on official IANA website do not have an extension provided, for example it's impossible to do web scraping from this page https://www.iana.org/assignments/media-types/application/msword as it has no mention of `.doc` extension being associated with `/msword` mime type.
- Read more on how browsers handle missing/generic mime types: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_Types#mime_sniffing
- Mozilla states that magic numbers aren't either a 100% valid solution: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_Types#other_methods_of_conveying_document_type
- Some magic numbers can be problematic, because for example a `.docx` file has the same magic number as `.zip` file (https://www.iana.org/assignments/media-types/application/vnd.openxmlformats-officedocument.wordprocessingml.document) - in this case the mime type is a better choice to determine file type that magic number
- Seems like even the main browsers have problems with determining types of some files. The strategy they use is described [here](https://stackoverflow.com/a/26303098) and can be summed as `best-effort` strategy that isn't 100% certain in all situations

Conclusion:
- With all I said above, it's unlikely we will find a solution that works 100% of the time - we will need to do assumptions about the file type in some kind of situations
- What we can do: check the mime type and if it's one of the generic mime types, assume that the file extension is the source of truth

**Comment by MariaHCD** [comment]
> if the mime type and the extension doesn't match they simply send the file as binary file without preview even if it have image extension

I think this might be potential solution. The backend validation of file mime types doesn't look to be working 100% either: https://github.com/Expensify/App/issues/21795

cc: @youssef-lr

**Comment by MariaHCD** [comment]
As an update here, we've removed the backend validation that tries to detect if the file is invalid by checking for it's mime type as the solution didn't work 100% of the time and it started blocking users from uploading valid files. For the same reason, we're also thinking that mime checking on the frontend is not required either.

A potential solution that @youssef-lr is looking into is using `RNImage`'s onError callback to disable the image preview if the file isn't actually a valid image. So we will continue to allow the file to be uploaded in the backend but we just won't render any preview for it. Is that right, @youssef-lr?

I think @MaciejSWM had a similar [solution](https://github.com/Expensify/App/issues/19728#issuecomment-1625310005) in mind:

> onError handler when error is thrown in <ImageView />

**Comment by KALLL** [comment]
As the issue reporter, I proposed the current solution's approach first in the expected result section on https://github.com/Expensify/App/issues/19728#issue-1729667192
"just upload and send as a file without trying to display preview(if the image file is invalid)"
And also I was explaining why this solution is the most efficient solution on the slack conversation https://expensify.slack.com/archives/C01GTK53T8Q/p1686934136190529?thread_ts=1686912137.101449&cid=C01GTK53T8Q

**Comment by MariaHCD** [comment]
Okay, thanks but regardless of who proposed the idea first, I think it was good for us to rule out other potential solutions.

Now we can focus on the solution that looks to be more promising: allow invalid files to be uploaded but don't render any preview for it. The backend already supports this since we reverted the previous PR that checks for mime types. So it looks to me that the next piece is to have the App support not rendering file previews for invalid files.

Waiting on @youssef-lr: are we thinking we want to evaluate the existing proposals (or accept new ones) for this or if this should go internal.

**Comment by youssef-lr** [comment]
@MariaHCD I had the code ready so went ahead with the PR. I've proposed this solution before [here](https://expensify.slack.com/archives/C049HHMV9SM/p1685001348461059) but back then we decided to :donothing:.

**Comment by MariaHCD** [comment]
Great, thanks for confirming! Taking this internal.

**Comment by MariaHCD** [comment]
@aimane-chnaif is reviewing the PR fix: https://github.com/Expensify/App/pull/23103

**Comment by KALLL** [comment]
Yes and FYI this issue is now over 2 months since created.

**Comment by youssef-lr** [comment]
PR pending review from @aimane-chnaif

**Comment by KALLL** [comment]
May I know what is this issue's progress, if it being done internally? @MariaHCD @aimane-chnaif @sonialiap

**Comment by KALLL** [comment]
Receiving new proposals? @youssef-lr @sonialiap

**Comment by KALLL** [comment]
So, is this task being progressed internally? I mean https://github.com/Expensify/App/issues/19728#issuecomment-1672079068 after this I thought this issue was waiting for PR review from @aimane-chnaif . So do we have an update?

**Comment by sonialiap** [comment]
@youssef-lr I am not sure how to check, has this PR been opened to contributors? https://github.com/Expensify/App/pull/23103#discussion_r1306693836

**Comment by youssef-lr** [comment]
Sorry, I'm working on higher priority issues. But I should pick this up again either this week or the next one.

**Comment by KALLL** [comment]
@aimane-chnaif @kbecciv  @sonialiap @mallenexpensify @MariaHCD Glad to see this issue resolved. I reported this issue, I am waiting to be paid as the issue reporter, thanks for your consideration.

**Comment by aimane-chnaif** [comment]
@KALLL We ended up not fixing as not a priority
